### ISSUE TITLE
Find a stable way to distribute binary in future

### ISSUE THREAD

**Issue by FranklinYu** [MacDownApp/macdown#1127]
We have temporarily fixed #1106 by manually removing the code signature. There are three potential long-term solutions:

1. Continue removing the signature with `codesign`. Not really sustainable since this is private API.
2. Create a second target, as mentioned in https://github.com/MacDownApp/macdown/issues/1106#issuecomment-562519332.
3. Create a company to hold the Developer Program account. Obstacles are mentioned in https://github.com/MacDownApp/macdown/issues/1106#issuecomment-558859778

**Comment by fadookie** [comment]
I have an inactive apple developer account under my personal name. I'd be willing to sign if we can raise funds to cover the annual membership costs. I can chip in $20 or so this year. However I don't know what the long-term effects of tying it to my personal account would be. Presumably if someone else needed to take it over in the future it might cause problems since I don't know how Sparkle would handle the public key changing.

**Comment by FranklinYu** [comment]
Yes, that can be the 4<sup>th</sup> way. @tempelmann offered similarly in https://github.com/MacDownApp/macdown/issues/1106#issuecomment-558694309. I’m not sure what user would think; I guess a poll is necessary before we actually go through this path.

**Comment by tempelmann** [comment]
You should not have to worry about Sparkle, as long as the private key is passed to the next maintainer who makes the releases. Just do not include the Sparkle key here in the repository.
However, if you have to make a new key, then Sparkle won't let users update older versions to the differently signed newer app.

For the regular code signing and notarization it does not matter who does it. Every new version could be signed by a different Apple Account, as Sparkle does not check that (or, if it does, you can easily disable that extra check). What should not change is the bundle ID - if that changes, Sparkle will not update the app (but there is no reason to ever change the bundle ID, as it's not tied to signing, so this should not be an issue).

So, make sure the Sparkle signing key is kept secure (and private) and can be passed to the next maintainer, and let anyone sign the apps for release. You don't have to have your own Account for that.

The only catch is that person who signs it with their account will be the one getting in trouble if Apple decides that the app (i.e. MacDown) is doing bad (illegal) stuff and wants to punish the developer. But as long as the app is open source, this shouldn't be a big risk. I'd personally not have a problem with offering the apps with my Apple account.

**Comment by FranklinYu** [comment]
Sparkle isn’t a problem at all. The safest solution is

1. Someone notarize the build and pass it to me (or post it here, it doesn’t contain secret).
2. I sign it with the Sparkle private key. This produces the signature (some Base64 string) without modifying the build.
3. I post the build as usual.

This way no private key need to be shared. As someone with security background I would try to avoid sharing any private key on the air.

**Comment by tempelmann** [comment]
Cool.

I'm codesigning and notarizing it now. Had to manually fix the sparkle framework first, though (see #1130).

Does notarization require some special access permissions? Such as for Calendar, Address book, Apple Events, etc? Does it generate code or load executable plugins?

**Comment by tempelmann** [comment]
Here's the signed and notarized app: 

Ideally, you should verify that the executables in that archive are the same that you have, to make sure I didn't build a "bad" version. Not sure how to compare that, though. Maybe you can remove the code signing again with the signing tools and then see if the files are identical to yours. Hopefully they are. Or instead codesign both the original and my version with an ad-hoc cert, so that both should have the same, and then compare them.

**Comment by fadookie** [comment]
I think if you extract the bundle and do a directory diff you could verify all the resources are the same - i think the signature is just a regular file inside the bundle. However, it's likely the darwin binary is different - i guess it's possible it's the same if it was build on the exact same Xcode version with the same optimization settings, etc. But I'm not sure if there is a 1:1 guarantee even then.

**Comment by FranklinYu** [comment]
Note that currently I’m not prioritizing this issue since Sparkle should be working. When a user update through Sparkle, another right-click isn’t required. (At least this is what I expect right now.) I created issue out of concern that Apple would further restrict unsigned application.

**Comment by FranklinYu** [comment]
By the way, I finally found out why this suddenly broke. It’s not anything we did, but [Apple disabled unsigned build since Xcode 10](https://stackoverflow.com/questions/31039513/how-can-i-skip-code-signing-for-development-builds-in-xcode#comment93808358_31985749). I’m trying [this hack to turn it off forcefully](https://stackoverflow.com/questions/31039513/how-can-i-skip-code-signing-for-development-builds-in-xcode#54296008) (thank you @mayoff!) in 22654227f3efe88fe6481c34f377a97be16fb612. I’m not sure whether this is less hacky than current `codesign` one or not, but hopefully it solves #1144. Let’s see how it goes.
